design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+--   in particular CREATE TABLE would create an empty table, and the AS part would load the query result into the new table;
+--   we might add the persistent query version only in the future)
+-- a KLIP-8 "pull query" (we can add filters, projections etc)
+-- the query conceptually "scans" the full inputTable (in particular the current generation when the query is issued),
+   computes the finite result, and terminates
+```
+
+
+**Example 250: a table-table join**
+
+```sql
+[CREATE TABLE myRdbmsStyleTable AS] SELECT * FROM inputTable t1 JOIN otherTable t2 ON t1.key = t2.key
+-- returns a TABLE
+-- same properties as Example 200
+-- because all inputs are a table, the output is a table
+```
+
+The last point is a design principle (cf. P2 above)
+
+* (P3) a query with only TABLES as inputs, returns a TABLE as output
+* (P3-a) we can extend this principle to n-ary TABLE joins
+
+An open question is, what table version/generations are joined to each other for this case – for unary queries,
+we can just use the "current generation of the current version" (i.e., whatever is in RocksDB atm when the query is issued).
+For n-ary queries, we need to figure out what semantics we want to have/allow (if we ever support this type of join to begin
+with—--currently, I would advocate to not allow this query at all, but tell users to define a continuous table-table join query with proper
+event-time semantics and query the result TABLE/MV instead).
+
+
+**Example 300: creating a materialized view (persistent continuous query of an input table)**
+
+```sql
+CREATE MATERIALIZED VIEW resultView AS SELECT * FROM inputTable
+-- returns a TABLE
+-- this is always a persistent query (without the CREATE MV AS prefix, the query is the same as Example 200 – compare Example 400 for the transient version)
+-- the result is a MV, ie, a table with a continuous queries that update the table when the base data changes; insert/update/delete statements are not allowed
+-- the same as current CTAS; for backward compatibility, current CTAS would still be supported but would be deprecated
+```
+
+
+**Example 400: subscribing to a changelog stream (transient continuous query of an input table)**
+
+```sql
+[CREATE STREAM myChangelog AS] SELECT * FROM STREAM(tableOrMaterializedView)
+-- returns a STREAM
+-- compare Example 100 (those properties apply)
+-- without CSAS prefix, it is the transient query counterpart to Example 300 – all updates are "streamed" continuously to the client
+--   (note, that the result is a fact-stream and it's the client responsibility to "interpret" the records with update semantics)
+-- The STREAM(...) operator creates a changelog stream from its input TABLE (can be a RDBMS style table or a MV); compare Example 0
+-- The output stream contains one record for each record in the table (ie, full table scan) based on the table generation when the query was issued
+--   plus a record for all future updates
+```
```

Comment:
I'm assuming you can do something like `STREAM(SELECT * FROM TABLE AS OF 0)` to get the full (compacted) change log of a table + updates?

Reply:
The full syntax would be
```
SELECT * FROM STREAM(SELECT * FROM TABLE AS OF 0)
```
It's an open question that I raised on the KLIP already, if `STREAM(...)` would be a valid query by itself from a grammar point of view? If yes, than `STREAM(SELECT * FROM TABLE AS OF 0)` would be a short version of the full syntax -- I don't have a strong opinion on this question about it atm (also not sure if this is a fundamental design question or a question about syntactic sugar?)